libdrgn/orc_info.c

```diff
@@ -367,18 +526,40 @@ struct drgn_error *drgn_module_parse_orc(struct drgn_module *module)
 {
 	struct drgn_error *err;
 
-	if (module->debug_file->platform.arch->arch != DRGN_ARCH_X86_64)
+	if (module->prog->platform.arch->arch != DRGN_ARCH_X86_64)
 		return NULL;
 
-	// pc_offsets and entries point to the Elf_Data buffers until we're
-	// done. We don't want those freed by drgn_module_orc_info_deinit(), so
-	// clear them if anything goes wrong.
+	// Depending on the source of pc_offsets and entries, we may or may not
+	// need to free them. We should always clear the pointers out of the
+	// module to avoid an invalid or double free in
+	// drgn_module_orc_info_deinit().
 	_cleanup_(drgn_module_clear_orc) struct drgn_module *clear = module;
+	_cleanup_free_ void *cleanup_pc_offsets = NULL;
+	_cleanup_free_ void *cleanup_entries = NULL;
+
+	if (module->debug_file) {
```

Comment:
If `debug_file` has (for some cursed reason) a different platform than the program, then `can_use_debug_file` in `drgn_module_parse_orc()` will be false, and it will call `drgn_module_parse_orc()` expecting to read the built-in ORC info. But then this will try to parse it from the debug file.

I think it'd be better to pass a `bool use_builtin` or something like that to `drgn_module_parse_orc()` to avoid that confusion.